src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -159,16 +161,25 @@ public DynamicObject group(DynamicObject thread) {
 
         @TruffleBoundary
         @Specialization
-        public DynamicObject kill(final DynamicObject rubyThread) {
-            final Thread toKill = Layouts.THREAD.getThread(rubyThread);
-
-            if (toKill == null) {
-                // Already dead
-                return rubyThread;
-            }
-
-            getContext().getSafepointManager().pauseThreadAndExecuteLater(toKill, this,
-                    (currentThread, currentNode) -> ThreadManager.exit(getContext(), currentThread, currentNode));
+        public DynamicObject kill(DynamicObject rubyThread) {
+            final ThreadManager threadManager = getContext().getThreadManager();
+            final DynamicObject rootThread = threadManager.getRootThread();
+
+            getContext().getSafepointManager().pauseAllThreadsAndExecute(this, false, (thread, currentNode) -> {
+                if (thread == rubyThread) {
+                    final FiberManager fiberManager = Layouts.THREAD.getFiberManager(thread);
+                    final DynamicObject fiber = fiberManager.getRubyFiberFromCurrentJavaThread();
+
+                    if (fiberManager.getCurrentFiber() == fiber) {
+                        if (thread == rootThread) {
+                            throw new RaiseException(coreExceptions().systemExit(0, currentNode));
```

Comment:
Maybe it doesn't matter because we're exiting, but this should probably require adding `throwsControlFlowException` to the `@TruffleBoundary`.

Reply:
It will only throw from this method if the thread is killing itself like `Thread.current.kill`.
In that case we don't need a safepoint, so throwsControlFlowException=true would avoid a deopt indeed, good catch.